### PR TITLE
Update to latest Oculus SDK, add input device support

### PR DESCRIPTION
**Pull request by jherico**
Migrate to Oculus consumer release SDK.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73bacb5aeb1b16c5c0831b7e1e87b897583a6249 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by hyperlogic** [comment]
Do you think it's worth while to update the ViveControllerManager to use the new LG and RG buttons?

**Comment by hyperlogic** [comment]
Currently we map Vive grip buttons to the left and right bumpers.  See ViveControllerManager::InputDevice::handleButtonEvent()

**Comment by ZappoMan** [comment]
I've removed development approved until we can sort out the behavior of the old runtimes with this build

**Comment by hyperlogic** [comment]
Somewhat leading question...  Now that 1.3 does automatic async time warp, do we need a present thread at all?

**Comment by ZappoMan** [comment]
we are not getting rid of the present thread 1 week before MVP.

**Comment by jherico** [comment]
> Currently we map Vive grip buttons to the left and right bumpers.

Hmmm... yeah, that's better.

**Comment by jherico** [comment]
@ZappoMan ok, this build should automatically disable the plugin if it sees the 0.8 runtime DLL.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 516e1ef94dc25f4dbf30a52f21e9b9f68dd0e7bb available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

File: plugins/oculus/src/OculusHelpers.cpp [hunk 1]
```diff
@@ -9,20 +9,34 @@
...
 bool oculusAvailable() {
     ovrDetectResult detect = ovr_Detect(0);
-    return (detect.IsOculusServiceRunning && detect.IsOculusHMDConnected);
+    if (!detect.IsOculusServiceRunning || !detect.IsOculusHMDConnected) {
+        return false;
+    }
+
+    // HACK Explicitly check for the presence of the 0.8 runtime DLL, and fail if it 
+    // exists (the ovr_Detect doesn't distinguish between 0.8 and 1.3)
+    if (QFile(BAD_OCULUS_RUNTIME_FILE).exists()) {
```
Comment (ZappoMan) [inline plugins/oculus/src/OculusHelpers.cpp]: would it be reasonable to report some kind of an error here? maybe to the log file? maybe to the user in a message box?

Comment (ZappoMan) [inline plugins/oculus/src/OculusHelpers.cpp]: note: this technique doesn't work on my system... my `LibOVRRT64_0_8.dll` is in `C:\Windows\System32`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35d20f90c45e8b178476cd192eddfbd05a612590 available at: https://gist.github.com/2cf5d5684bb56a45e1fc<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

File: plugins/oculus/src/OculusHelpers.cpp [hunk 2]
```diff
@@ -9,20 +9,34 @@
...
     ovrDetectResult detect = ovr_Detect(0);
-    return (detect.IsOculusServiceRunning && detect.IsOculusHMDConnected);
+    if (!detect.IsOculusServiceRunning || !detect.IsOculusHMDConnected) {
+        return false;
+    }
+
+    // HACK Explicitly check for the presence of the 1.0 runtime DLL, and fail if it 
+    // doesn't exist
+    if (!QFile(GOOD_OCULUS_RUNTIME_FILE).exists()) {
+        return false;
```
Comment (ZappoMan) [inline plugins/oculus/src/OculusHelpers.cpp]: I still think you should add a log message that say something about "1.0 runtime not detected" --- so that if there's some problem with this in the future, we'll have an idea why it's not adding oculus to the menu

**Comment by ZappoMan** [comment]
This works for me... specifically, on my machine with 0.8 runtime installed, it does not show the oculus menu on the display menu

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3208865ef3acb7bc3ad0b518fbcb868d2faa8815 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by ZappoMan** [comment]
Test plan:
- smoke test to make sure nothing broke
- run with 0.8 runtime - you should see a message in your log file... and the oculus menu will not be present
- run with the 1.0 runtime - you should get the oculus menu -- and it should work correctly

**Comment by cozza13** [comment]
testing

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3616976350b3fa0b6d0b1705a0a5f3a8d33a8938 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 694a9942306554aa566999a4e4e5307632a30218 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by TNS-TrueNorth** [comment]
Testing PR7491 (for commit 694a994)
Pass - Running with 0.8 runtime, the correct Warning message appears "[WARNING] Oculus Runtime detected, but no 1.x DLL present" and the menu option for Oculus is not available.
Will test with 1.0 runtime soon

**Comment by jherico** [comment]
@TNS-TrueNorth how did you set up the 1.3 runtime with the DK2?  It should work.

**Comment by judasshuffle** [comment]
Works for me with the dk2
But in the rift config thing I went to settings general and enabled unknown sources.I'm not sure the rift appears in the controllers without that checked.
It does tell me the dk2 is unsupported 
I also had a rift install bug which this was the cure, in case it crops up for anyone else
https://www.reddit.com/r/oculus/comments/4caiwr/failed_install/

**Comment by hifi-gustavo** [comment]
Build console output for commit: b799b59b4cfd801ee5eb172528d2a2d8a01184db available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by TNS-TrueNorth** [comment]
@jherico Yesterday, we installed Oculus' customer facing setup software from https://www.oculus.com/en-us/setup/    The Oculus setup software is refusing to acknowledge the DK2 (screenshot attached).  But when the user skips past setup, the DK2 and Sensor appear on the Settings/Devices page even though the Oculus store says that the DK2 is not supported (screenshot attached).
oculus setup 2016-03-28 does not recognize dk2 hdmi usb or sensor usb
dk2 plugged in but not supported

**Comment by cozza13** [comment]
I tested this on CV1 and it performed well. I had 90 on demo and playa

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7c89d2361603c955f6cd3f6a95e1d5383d71be60 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5e26459d025b633ca55bc2cacdd9b3cde6b1f02a available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 605511576455e6d392f8da625e7e3d867d5778a9 available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f569cbf70ef40c15699acb31ef374e5b31c25cee available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by jherico** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: f569cbf70ef40c15699acb31ef374e5b31c25cee available at: https://gist.github.com/cd2ff674cd6afa884940<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7491.dmg

**Comment by cozza13** [comment]
Tested on Dk2 and it worked. 
Tested on CV1 on Playa and i was getting 90's

File: plugins/oculus/src/OculusBaseDisplayPlugin.cpp [hunk 3]
```diff
@@ -42,36 +47,30 @@ bool OculusBaseDisplayPlugin::internalActivate() {
 
     _hmdDesc = ovr_GetHmdDesc(_session);
 
-    _ipd = ovr_GetFloat(_session, OVR_KEY_IPD, _ipd);
-
     glm::uvec2 eyeSizes[2];
     _viewScaleDesc.HmdSpaceToWorldScaleInMeters = 1.0f;
 
+    _ipd = 0;
```
Comment (ZappoMan) [inline plugins/oculus/src/OculusBaseDisplayPlugin.cpp]: why did you change this?

**Comment by ZappoMan** [comment]
Passes my code review.
I am going to upgrade to the 1.3 runtime now, and do some testing.

File: plugins/oculus/src/OculusBaseDisplayPlugin.cpp (hunk 3 above)
Comment (jherico) [inline plugins/oculus/src/OculusBaseDisplayPlugin.cpp]: the oculus SDK no longer exposes a variable to fetch the IPD directly.  Instead it just applies it to the eye poses and we have to calculate it for ourselves.

**Comment by ZappoMan** [comment]
Bugs I've found in testing:
- Crash if you close the oculus window, while interface is running, and then try to switch to oculus view
https://www.bugsplatsoftware.com/individualCrash/?id=2454&database=interface_alpha

**Comment by RyanDowne** [comment]
I am seeing a lot of avatar judder and overlay judder when walking and not seeing it on master.

To repro put on the Art3mis avatar in the marketplace. 
Go to demo
Walk outside with your hands in front of you and look at them
The overlays that judder are of the loading bar variety. I don't see the edit button judder.

**Comment by ZappoMan** [comment]
@jherico - I can confirm that what @RyanDowne is seeing is definitely happening in this PR.

**Comment by jherico** [comment]
I can't test the progress bar in production because the effect would be specific to the fact that in the HMD the progress bar is a 3D overlay, not a 2D overlay (although we could clearly fix it by changing it to be a 2D overlay always).

However, I'm definitely seeing the hand judder described by @RyanDowne in production.  if I go into third person and fly my avatar around, it's very easy to see the issue, becuase my hand will lag behind on a vector which is opposite the direction of travel and longer the faster I'm moving.  this is specifically happening to the avatar hands, not to the avatar as a whole.

the judder appears to be that the number of frames behind is actually variable, which leads me to suspect that it's actually being updated asynchronously relative to rendering.
